handler.go

```diff
@@ -132,153 +130,137 @@ func (ca *Cagent) PostResultsToHub(result Result) error {
+		errs = append(errs, err.Error())
+	}
 
-		results.Measurements = results.Measurements.AddWithPrefix("system.", ipResults)
+	measurements = measurements.AddWithPrefix("proc.", proc)
 
-		fsResults, err := fs.Results()
-		if err != nil {
-			// no need to log because already done inside fs.Results()
-			errs = append(errs, err.Error())
-		}
+	mem, err := ca.MemResults()
+	if err != nil {
+		// no need to log because already done inside MemResults()
+		errs = append(errs, err.Error())
+	}
 
-		results.Measurements = results.Measurements.AddWithPrefix("fs.", fsResults)
+	measurements = measurements.AddWithPrefix("mem.", mem)
 
-		netResults, err := net.Results()
-		if err != nil {
-			// no need to log because already done inside net.Results()
-			errs = append(errs, err.Error())
-		}
+	swap, err := ca.SwapResults()
+	if err != nil {
+		// no need to log because already done inside MemResults()
+		errs = append(errs, err.Error())
+	}
 
-		results.Measurements = results.Measurements.AddWithPrefix("net.", netResults)
+	measurements = measurements.AddWithPrefix("swap.", swap)
 
-		proc, err := ca.ProcessesResult()
+	if runtime.GOOS == "linux" {
+		raid, err := ca.RaidState()
 		if err != nil {
-			// no need to log because already done inside ProcessesResult()
+			// no need to log because already done inside RaidState()
 			errs = append(errs, err.Error())
 		}
 
-		results.Measurements = results.Measurements.AddWithPrefix("proc.", proc)
+		measurements = measurements.AddWithPrefix("raid.", raid)
+	}
 
-		mem, err := ca.MemResults()
+	if runtime.GOOS == "windows" {
+		wu, err := ca.WindowsUpdatesWatcher().WindowsUpdates()
 		if err != nil {
 			// no need to log because already done inside MemResults()
 			errs = append(errs, err.Error())
 		}
 
-		results.Measurements = results.Measurements.AddWithPrefix("mem.", mem)
+		measurements = measurements.AddWithPrefix("windows_update.", wu)
+	}
 
-		swap, err := ca.SwapResults()
-		if err != nil {
-			// no need to log because already done inside MemResults()
-			errs = append(errs, err.Error())
-		}
+	if len(errs) == 0 {
+		measurements["cagent.success"] = 1
+		return measurements, nil
+	}
 
-		results.Measurements = results.Measurements.AddWithPrefix("swap.", swap)
+	measurements["cagent.success"] = 0
 
-		if runtime.GOOS == "linux" {
-			raid, err := ca.RaidState()
-			if err != nil {
-				// no need to log because already done inside RaidState()
-				errs = append(errs, err.Error())
-			}
+	return measurements, errors.New(strings.Join(errs, "; "))
+}
 
-			results.Measurements = results.Measurements.AddWithPrefix("raid.", raid)
-		}
+func (ca *Cagent) ReportMeasurements(measurements MeasurementsMap, outputFile *os.File) error {
+	result := Result{Timestamp: time.Now().Unix(), Measurements: measurements}
 
-		if runtime.GOOS == "windows" {
-			wu, err := wuw.WindowsUpdates()
+	if outputFile != nil {
+		jsonEncoder := json.NewEncoder(outputFile)
+		err := jsonEncoder.Encode(result)
```

Comment:
Don't you have to pass a reference here?

Reply:
not critical, but fixed

Reply:
Does it even work if you _don't_ pass a reference?
Usually things are passed by value, so they're copied. Therefor if you don't pass a reference the `Encode` function can't modify the struct you passed 🤔

Reply:
it doesn't need to modify. It is encoder, not decoder :)

Reply:
Gotcha... true 🙈